### PR TITLE
Enhanced RDoc for IO.popen

### PR DESCRIPTION
**Pull request by BurdetteLamar**
This is tough.

### CODE REVIEW

**Review by jeremyevans — CHANGES_REQUESTED** [review]
Looks mostly good, please see requested changes.

File: io.c [hunk 1]
```diff
@@ -7366,35 +7366,87 @@ static VALUE popen_finish(VALUE port, VALUE klass);
...
+ *  Executes the given command +cmd+ as a subprocess
+ *  whose $stdin and $stdout are connected to a new strean +io+.
  *
- *  The PID of the started process can be obtained by IO#pid method.
+ *  Argument +cmd+ is one of the following (each detailed below):
  *
- *  _cmd_ is a string or an array as follows.
+ *  - The string <tt>'-'</tt>: Forks a new instance of Ruby as a subprocess.
+ *  - Any other string: Passes the string to a shell as a command.
+ *  - An array of strings: Runs a Ruby subprocess with the array as its ARGV.
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply io.c:7379]: @jeremyevans, you're referring to this?
-   `[env, [cmdname, argv0], arg1, ..., opts] : command name, argv[0] and zero or more arguments (no shell)`
I don't understand ths form.  Wouldn't arg1 have to be mode?  And what does the ellipsis mean?

[the remark this answers is not in the document]
Comment (jeremyevans) [reply io.c:7379]: I think the previous code example is wrong.  You can pass arguments without using a shell using an array: `IO.popen(['cmd', 'arg1', 'arg2'])`.  You can set argv0 by making the first element of this array be an array with the command and desired argv0: `IO.popen([['cmd', 'argv0'], 'arg1', 'arg2'])`

[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply io.c:7379]: Ok, I have this:

File.write('t.rb', '$stderr.puts ARGV')# => 17
IO.popen([['ruby', 'argv0'], 't.rb', 'FOO', 'BAR'])# => #<IO:fd 5>
FOO
BAR

What is `argv0` doing?

[the remark this answers is not in the document]
Comment (jeremyevans) [reply io.c:7379]: `argv0` sets `argv[0]` when running the program.  Some programs access `argv[0]` and operate differently depending on the value.  For example, on my system, `tar` and `pax` are the same program, and whether it operates as `tar` or `pax` depends on `argv[0]`.

[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply io.c:7379]: I'm still missing something:

File.write('t.rb', '$stderr.puts argv[0]') # => 20
IO.popen([['ruby', 'argv0'], 't.rb', 'FOO', 'BAR']) # => #<IO:fd 5>
t.rb:1:in `<main>': undefined local variable or method `argv' for main:Object (NameError)

[the remark this answers is not in the document]
Comment (jeremyevans) [reply io.c:7379]: I don't think Ruby provides access to `argv[0]`, so it isn't a good program to test with.  Try using a shell:

```
$ ruby -e 'puts IO.popen(["/bin/sh", "-c", "echo $0"]).read'
/bin/sh
$ ruby -e 'puts IO.popen([["/bin/sh", "foo"], "-c", "echo $0"]).read'
foo
```

[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply io.c]: Fixed?

File: io.c [hunk 2]
```diff
@@ -7366,35 +7366,87 @@ static VALUE popen_finish(VALUE port, VALUE klass);
...
+ *      pipe.puts('ls')
+ *      pipe.close_write
+ *      $stderr.puts pipe.readlines.size
+ *    end
+ *
+ *  Output:
+ *
+ *    213
+ *
+ *  <b>Ruby Subprocess</b>
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply io.c]: Fixed.

File: io.c [hunk 3]
```diff
@@ -7412,22 +7464,6 @@ static VALUE popen_finish(VALUE port, VALUE klass);
  *      ls_result_with_error = ls_io.read
  *    }
  *
- *  Raises exceptions which IO.pipe and Kernel.spawn raise.
- *
- *  If a block is given, Ruby will run the command as a child connected
- *  to Ruby with a pipe. Ruby's end of the pipe will be passed as a
- *  parameter to the block.
- *  At the end of block, Ruby closes the pipe and sets <code>$?</code>.
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply io.c:7420]: Now mentioned near the top.

File: io.c [hunk 4]
```diff
@@ -7366,35 +7366,87 @@ static VALUE popen_finish(VALUE port, VALUE klass);
...
+ *  Optional argument +mode+ may be any valid \IO mode.
+ *  See {Modes}[#class-IO-label-Modes].
  *
- *  If _cmd_ is a +String+ ``<code>-</code>'',
- *  then a new instance of Ruby is started as the subprocess.
+ *  The optional hash argument +env+ specifies name/value pairs that are to be added
+ *  to the environment variables for the subprocess:
  *
- *  If <i>cmd</i> is an +Array+ of +String+,
- *  then it will be used as the subprocess's +argv+ bypassing a shell.
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply io.c:7391]: Fixed?

**Review by jeremyevans — APPROVED** [review]
Looks good!

**Pull request merged**